Game of Life MVP

CimimUxMaio opened the pull request:
The MVP includes:
- Running the game's rules on a window using *macroquad*.
- 2 operation modes: "Editing" mode and "Running" mode.
    - The ability to edit the current state of the simulation using left-click.
    - Keybinding for switching between modes (space bar).
- Simple text UI indicating current mode and a short help message.
- README.md file with documentation on how to set it up and how to play.

Commit:
Implement `Grid` structure

Implemented Grid structure.
Includes method functions for:
- Drawing the grid.
- Updating cells in the grid.

Commit:
Update formating errors

Commit:
Add tests for `Grid` structure

Test detected that Grid::new() was generating a grid where all
cells where `true`. That is updated and now a new grid contains
all `false` values.

Commit:
Merge pull request #3 from CimimUxMaio/feature-grid-model

Implement `Grid` structure

Commit:
Extend `Grid` model

- Updated matrix initialization as suggested in #3.
- Defined new methods:
    - coordinates/0
    - neighbours_state/0
- Defined a `Clone` implementation.

Commit:
Implement game rules and update grid

- Implemented game rules in src/model/rules.
- Main program now updates de grid following the update rules.
- Framerate is limited for easier visualization.

Commit:
Merge pull request #4 from CimimUxMaio/grid-model-update

Extend `Grid` model

Commit:
Implement tests for `compute_new_state`

Commit:
Fix unexpected rule behaviour on update

In the update function (previously named "run") new state
was computed using the current grid instead of the previous.

Commit:
Fix neighbour computation for borders and corners

Corner and border cells now compute their neighbours correctly.
- Test were added for this cases.

Removed all logic related to wrapping moving cells when they
get out of boundaries.
- Now cells don't wrap arround the grid for simplicity.
- This means that `grid.get()` with invalid `x` and `y`
  values will fail.

Commit:
Update tests for overflowing positions

Now given coordinates should not overflow.
The expected behaviour is to panic when invalid coordinates
are given.

Commit:
Merge pull request #5 from CimimUxMaio/feature-update-rules

Feature update rules

Commit:
Update README.md with quick start and dependencies

Commit:
Merge pull request #6 from CimimUxMaio/update-docs

Update README.md with quick start and dependencies

Commit:
Add editing mode

Editing mode allows users to draw on the screen.

User can switch between editing and running modes by
pressing space bar.

UI text labels were added to indicate the mode the user
is currently on and a short help message at the bottom.

Commit:
Update README with guide and rules

- Added a guide section indicating how the user may navigate between
  each operating mode.
- Added a rules section, indicating the Game of Life's ruleset.

Commit:
Add image to README quick guide

Commit:
Merge 'draw-mode' feature into feature-tutorial

Commit:
Merge pull request #7 from CimimUxMaio/feature-draw-mode

Add Editing Mode

Commit:
Merge pull request #8 from CimimUxMaio/feature-tutorial

User guide

gabrielbosio commented on src/main.rs:
```diff
@@ -1,3 +1,77 @@
+    }
+}
+
+fn draw_ui(is_editing: &bool) {
```
```suggestion
fn draw_ui(is_editing: bool) {
```

Is not necessary to pass this boolean as a reference

CimimUxMaio replied:
It is updated.

gabrielbosio commented on src/main.rs:
```diff
@@ -1,3 +1,77 @@
+        let new_state = model::rules::compute_new_state(prev, x, y);
+        grid.update(x, y, new_state);
+    }
+}
```
```suggestion
fn update(grid: &mut model::Grid) {
    let prev = grid.clone();
    
    for (x, y) in prev.coordinates() {
        let new_state = model::rules::compute_new_state(prev, x, y);
        grid.update(x, y, new_state);
    }
}
```

The `prev` variable is only used here so you can declare in this scope instead of declaring it outside

CimimUxMaio replied:
Great idea, I updated it.

gabrielbosio commented on src/main.rs:
```diff
@@ -1,3 +1,77 @@
+            edit_mode(&mut grid);
+        } else {
+            update(&prev_grid, &mut grid);
+        }
```
I think the name of both functions are a bit unclear because both functions update the grid in different ways. It could be clearer if the name of each function describe what they are updating. For example:

```suggestion
        if is_editing {
            update_edit_mode(&mut grid);
        } else {
            update_execution_mode(&prev_grid, &mut grid);
        }
```

CimimUxMaio replied:
Agreed, I updated it.

gabrielbosio commented on src/model/grid.rs:
```diff
@@ -0,0 +1,217 @@
+    }
+}
+
+impl Clone for Grid {
+    fn clone(&self) -> Self {
+        Grid {
+            width: self.width,
+            height: self.height,
+            matrix: self.matrix.clone(),
+        }
+    }
+}
```
You can replace this `impl` block with a `derive` attribute in the `Grid` struct definition:

```rust
#[derive(Clone)]
pub struct Grid {
    pub width: usize,
    pub height: usize,
    matrix: Vec<Vec<bool>>,
}
```

CimimUxMaio replied:
Thanks, I missed that. It is updated now.

gabrielbosio commented on src/model/grid.rs:
```diff
@@ -0,0 +1,217 @@
+
+        for row_n in 0..self.width {
+            for col_n in 0..self.height {
+                let is_live = self.matrix[row_n][col_n];
```
```suggestion
                let is_alive = self.matrix[row_n][col_n];
```

CimimUxMaio replied:
Updated

gabrielbosio commented on src/model/grid.rs:
```diff
@@ -0,0 +1,217 @@
+                    continue;
+                }
+
+                // If live, draw cell.
```
```suggestion
                // If it is alive, draw cell.
```

CimimUxMaio replied:
Updated.

gabrielbosio commented on src/model/grid.rs:
```diff
@@ -0,0 +1,217 @@
+                let is_live = self.matrix[row_n][col_n];
+
+                if !is_live {
+                    // If not live don't draw.
```
```suggestion
                    // If it is not alive, don't draw.
```

CimimUxMaio replied:
Updated.

gabrielbosio commented on src/model/rules.rs:
```diff
@@ -0,0 +1,105 @@
+use super::grid::Grid;
+
+pub fn compute_new_state(grid: &Grid, x: usize, y: usize) -> bool {
+    let cell_is_live = grid.get(x, y);
```
```suggestion
    let cell_is_alive = grid.get(x, y);
```

or

```suggestion
    let is_cell_alive = grid.get(x, y);
```

CimimUxMaio replied:
Updated

gabrielbosio commented on src/model/rules.rs:
```diff
@@ -0,0 +1,105 @@
+use super::grid::Grid;
+
+pub fn compute_new_state(grid: &Grid, x: usize, y: usize) -> bool {
```
I think this function name is not very clear because `compute_new_state` seems to imply computing the state of the entire grid:

```suggestion
pub fn compute_cell_new_state(grid: &Grid, x: usize, y: usize) -> bool {
```

CimimUxMaio replied:
I agree, I updated it.

gabrielbosio reviewed (dismissed):
Looks good in general. I requested some specific changes. Also I would add the following:

- Error handling
- Code documentation

Commit:
Remove update_ui parameter as reference

Commit:
Simplify main() and update()

`prev_grid` was only used inside this function, so its update logic
was moved inside it. Simplifying the main loop.

Commit:
Rename update functions for different modes

- `update` was renamed to `update_execution_mode`.
- `edit_mode` was renamed to `update_edit_mode`.

Commit:
Replace Clone implementation for Grid with `derive`

Commit:
Update is_live to is_alive

Commit:
Rename `compute_new_state` function

Renamed `compute_new_state` to `compute_cell_new_state`.

Commit:
Document all functions

CimimUxMaio commented:
I've updated the code based on the comments and, in my last commit, I added documentation for all functions.

santiago-MV reviewed (approved):
Approved!
